File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
n = int(input())
s = [ int(input()) for _ in range(n) ]
d = [-1] * n

d[0] = s[0]
if n > 1:
    d[1] = s[0] + s[1]
if n > 2:
    d[2] = max(s[0], s[1]) + s[2]

for i in range(3, n):
    d[i] = max(d[i-2], s[i-1] + d[i-3]) + s[i]

print(d[n-1])

File: 백준/Silver/2579. 계단 오르기/계단 오르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    // 계단 오르기
    // 실버 3

    static Integer[] dp;
    static int[] arr;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        dp = new Integer[n];
        arr = new int[n];

        for(int i=0;i<n;i++){
            arr[i] = Integer.parseInt(br.readLine());
        }
        dp[0] = arr[0];
        if(n > 1) {
            dp[1] = arr[0] + arr[1];
        }
        System.out.println(step(n-1));
    }

    private static int step(int i){
        if (i<0){
            return 0;
        }
        // i번째 계단을 무조건 밟는 선에서 얻을 수 있는 최고 점수
        if(dp[i] == null){
            dp[i] = Math.max(step(i-3) + arr[i-1] + arr[i], step(i-2) + arr[i]);
        }

        return dp[i];
    }
}
